// wdio-html-nice-reporter/lib/htmlReporter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const reporter_1 = __importDefault(require("@wdio/reporter"));
const htmlGenerator_1 = __importDefault(require("./htmlGenerator"));
const types_1 = require("./types");
const dayjs_1 = __importDefault(require("dayjs"));
const wdio_report_events_1 = __importDefault(require("@rpii/wdio-report-events"));
const typescript_string_operations_1 = require("typescript-string-operations");
const fs = require('fs-extra');
const path = require('path');
const log4js = require('@log4js-node/log4js-api');
let reportProxy = new wdio_report_events_1.default();
const timeFormat = "YYYY-MM-DDTHH:mm:ss.SSS[Z]";
class HtmlReporter extends reporter_1.default {
    constructor(options) {
        super(Object.assign({
            stdout: true,
            logFile: './logs/reporter.log',
        }, options));
        this._suiteIndents = {};
        this._suiteUids = new Map();
        this._testUids = new Map();
        let opts = {
            stdout: true,
            outputDir: 'reports/html-reports/',
            filename: 'report.html',
            reportTitle: 'Test Report Title',
            useOnAfterCommandForScreenshot: false
        };
        this.options = Object.assign(opts, options);
        if (!this.options.LOG) {
            this.options.LOG = log4js.getLogger(this.options.debug ? 'debug' : 'default');
            ;
        }
        const dir = this.options.outputDir + 'screenshots';
        fs.ensureDirSync(dir);
        this._indents = 0;
        this._suiteIndents = {};
        this.metrics = new types_1.Metrics();
        this.openInProgress = false;
        this.defaultTestIndent = '   ';
        this._currentSuiteUid = "suite uid";
        this._currentTestUid = "test uid";
        this._currentCid = "cid";
        reportProxy.connectMessageEvent(this.saveMessage.bind(this));
        reportProxy.connectScreenshotEvent(this.saveScreenshot.bind(this));
        reportProxy.connectVideoCaptureEvent(this.saveVideo.bind(this));
    }
    get isSynchronised() {
        return !this.openInProgress;
    }
    onRunnerStart(runner) {
        this.options.LOG.info(typescript_string_operations_1.String.Format("onRunnerStart: {0}", runner.cid));
        this.options.LOG.debug(JSON.stringify(runner));
        //todo look at fix, not async safe. but one cid per report file
        this._currentCid = runner.cid;
        this.metrics.passed = 0;
        this.metrics.skipped = 0;
        this.metrics.failed = 0;
        this.metrics.start = (0, dayjs_1.default)().utc().format("YYYY-MM-DDTHH:mm:ss.SSS[Z]");
    }
    onSuiteStart(suite) {
        this._suiteUids.set(suite.uid, suite);
        if (suite.type === 'feature') {
            this._indents = 0;
            this._suiteIndents[suite.uid] = this._indents;
        }
        else {
            this._suiteIndents[suite.uid] = ++this._indents;
        }
        this._currentSuiteUid = suite.uid;
        suite.start = new Date();
        this.options.LOG.info(typescript_string_operations_1.String.Format("onSuiteStart: {0}:{1}", suite.cid, suite.uid));
        this.options.LOG.debug(JSON.stringify(suite));
    }
    onTestStart(theTest) {
        this.options.LOG.info(typescript_string_operations_1.String.Format("onTestStart: {0}:{1}", theTest.cid, theTest.uid));
        this.options.LOG.debug(JSON.stringify(theTest));
        this._currentTestUid = theTest.uid;
        //@ts-ignore
        theTest.events = [];
        //@ts-ignore
        theTest.errorIndex = 0;
        this._testUids.set(theTest.uid, theTest);
    }
    onTestPass(theTest) {
        this.options.LOG.info(typescript_string_operations_1.String.Format("onTestPass: {0}:{1}", theTest.cid, theTest.uid));
        this.options.LOG.debug(JSON.stringify(theTest));
        let test = this.getTest(theTest.uid);
        if (test) {
            this.moveErrorsToEvents(test);
        }
        this.metrics.passed++;
    }
    onTestSkip(test) {
        this.options.LOG.info(typescript_string_operations_1.String.Format("onTestSkip: {0}:{1}", test.cid, test.uid));
        this.options.LOG.debug(JSON.stringify(test));
        this.metrics.skipped++;
    }
    onTestFail(theTest) {
        this.options.LOG.info(typescript_string_operations_1.String.Format("onTestFail: {0}:{1}", theTest.cid, theTest.uid));
        this.options.LOG.debug(JSON.stringify(theTest));
        let test = this.getTest(theTest.uid);
        if (test) {
            this.moveErrorsToEvents(test);
        }
        this.metrics.failed++;
    }
    onTestEnd(theTest) {
        this.options.LOG.info(typescript_string_operations_1.String.Format("onTestEnd: {0}:{1}", theTest.cid, theTest.uid));
        this.options.LOG.debug(JSON.stringify(theTest));
        let test = this.getTest(theTest.uid);
        if (test) {
            this.moveErrorsToEvents(test);
        }
    }
    onHookStart(hook) {
        this.options.LOG.info(typescript_string_operations_1.String.Format("onHookStart: {0}:{1}", hook.cid, hook.uid));
    }
    onHookEnd(hook) {
        this.options.LOG.info(typescript_string_operations_1.String.Format("onHookEnd: {0}:{1}", hook.cid, hook.uid));
        if (hook.error) {
            this.metrics.failed++;
        }
    }
    onSuiteEnd(suite) {
        this.options.LOG.info(typescript_string_operations_1.String.Format("onSuiteEnd: {0}:{1}", suite.cid, suite.uid));
        this.options.LOG.debug(JSON.stringify(suite));
        this._indents--;
        suite.end = new Date();
        // this is to display suite end time and duration in master report.
        // for (const suiteInfo of this._suiteStats) {
        //     if (suiteInfo.uid == suite.uid) {
        //         suiteInfo.end = suite.end;
        //         suiteInfo._duration = suite._duration;
        //         break;
        //     }
        // }
    }
    isScreenshotCommand(command) {
        const isScreenshotEndpoint = /\/session\/[^/]*(\/element\/[^/]*)?\/screenshot/;
        return (
        // WebDriver protocol
        (command.endpoint && isScreenshotEndpoint.test(command.endpoint)) ||
            // DevTools protocol
            command.command === 'takeScreenshot');
    }
    //this is a hack to get around lack of onScreenshot event
    onAfterCommand(command) {
        if (this.options.useOnAfterCommandForScreenshot) {
            if (this.isScreenshotCommand(command) && command.result.value) {
                let timestamp = (0, dayjs_1.default)().format('YYYYMMDD-HHmmss.SSS');
                const filepath = path.join(this.options.outputDir, '/screenshots/', encodeURIComponent(this._currentCid), timestamp, this.options.filename + '.png');
                this.options.LOG.info(typescript_string_operations_1.String.Format("onAfterCommand: {0}:{1} taking screenshot {2}", this._currentCid, this._currentTestUid, filepath));
                fs.outputFileSync(filepath, Buffer.from(command.result.value, 'base64'));
                let test = this.getTest(this._currentTestUid);
                if (test) {
                    //@ts-ignore
                    test.events.push({ type: 'screenshot', value: filepath });
                }
            }
        }
    }
    onRunnerEnd(runner) {
        this.options.LOG.info(typescript_string_operations_1.String.Format("onRunnerEnd: {0}", runner.cid));
        this.options.LOG.debug(JSON.stringify(runner));
        this.openInProgress = true;
        this.metrics.end = (0, dayjs_1.default)().utc().format();
        this.metrics.duration = runner._duration;
        //error handling protection
        if (!this._currentSuiteUid) {
            this._currentSuiteUid = "suite";
        }
        if (!this._currentCid) {
            this._currentCid = "cid";
        }
        let suites = this.filterChildSuites();
        let reportFile = path.join(process.cwd(), this.options.outputDir, encodeURIComponent(this._currentSuiteUid), encodeURIComponent(this._currentCid), this.options.filename);
        let reportData = new types_1.ReportData(this.options.reportTitle, runner, suites, this.metrics, reportFile, this.options.browserName);
        htmlGenerator_1.default.htmlOutput(this.options, reportData, () => {
            this.openInProgress = false;
        });
    }
    getSuite(uid) {
        if (this._suiteUids.has(uid)) {
            return this._suiteUids.get(uid);
        }
        return undefined;
    }
    removeSuite(uid) {
        if (uid && this._suiteUids.has(uid)) {
            this._suiteUids.delete(uid);
        }
    }
    getTest(uid) {
        if (uid && this._testUids.has(uid)) {
            return this._testUids.get(uid);
        }
    }
    //this is a hack.  we have to move all the things in test.errors before they get blown away
    moveErrorsToEvents(test) {
        if (test.errors) {
            //@ts-ignore
            for (let i = test.errorIndex; i < test.errors.length; i++) {
                let errorObj = test.errors[i];
                let stack = test.errors[i].stack;
                if (stack && stack.includes("AssertionError")) {
                    errorObj = {
                        //@ts-ignore
                        message: test.errors[i].message.split("      \n").shift(),
                        stack: test.errors[i].stack,
                    };
                }
                //@ts-ignore
                test.events.push(new types_1.InternalReportEvent('Error', errorObj));
            }
            //@ts-ignore
            test.errorIndex = test.errors.length;
        }
    }
    saveScreenshot(filepath) {
        let test = this.getTest(this._currentTestUid);
        if (test) {
            this.moveErrorsToEvents(test);
            //@ts-ignore
            test.events.push(new types_1.InternalReportEvent('screenshot', filepath));
        }
    }
    saveVideo(filepath) {
        let test = this.getTest(this._currentTestUid);
        if (test) {
            this.moveErrorsToEvents(test);
            //@ts-ignore
            test.events.push(new types_1.InternalReportEvent('video-capture', filepath));
        }
    }
    saveMessage(message) {
        const test = this.getTest(this._currentTestUid);
        if (test) {
            this.moveErrorsToEvents(test);
            //@ts-ignore
            test.events.push({ type: 'log', value: message });
        }
    }
    filterChildSuites() {
        let suites = Array.from(this._suiteUids.values());
        for (let i = suites.length - 1; i >= 0; i--) {
            let parentSuite = suites[i];
            for (let k = parentSuite.suites.length - 1; k >= 0; k--) {
                let suite = parentSuite.suites[k];
                this.removeSuite(suite.uid);
            }
        }
        return Array.from(this._suiteUids.values());
    }
    indent(uid) {
        const indents = this._suiteIndents[uid];
        return indents === 0 ? '' : Array(indents).join('    ');
    }
}
exports.default = HtmlReporter;
